metrics/metrics.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2024 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"net/http"
+
+	cmlisters "github.com/cert-manager/cert-manager/pkg/client/listers/certmanager/v1"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	internalapiutil "github.com/cert-manager/csi-lib/internal/api/util"
+	"github.com/cert-manager/csi-lib/storage"
+)
+
+const (
+	// Namespace is the namespace for csi-lib metric names
+	namespace = "certmanager"
+	subsystem = "csi"
+)
+
+// Metrics is designed to be a shared object for updating the metrics exposed by csi-lib
+type Metrics struct {
+	log      logr.Logger
+	registry *prometheus.Registry
+
+	driverIssueCallCountTotal   *prometheus.CounterVec
+	driverIssueErrorCountTotal  *prometheus.CounterVec
+	certificateRequestCollector prometheus.Collector
+}
+
+// New creates a Metrics struct and populates it with prometheus metric types.
+func New(
+	nodeId string,
+	logger *logr.Logger,
+	registry *prometheus.Registry,
+	metadataReader storage.MetadataReader,
+	certificateRequestLister cmlisters.CertificateRequestLister,
+) *Metrics {
+	var (
+		// driverIssueCallCountTotal is a Prometheus counter for the number of issue() calls made by the driver.
+		driverIssueCallCountTotal = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "driver_issue_call_count_total",
+				Help:      "The number of issue() calls made by the driver.",
+			},
+			[]string{"node", "volume"},
+		)
+
+		// driverIssueErrorCountTotal is a Prometheus counter for the number of errors encountered
+		// during the driver issue() calls.
+		driverIssueErrorCountTotal = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "driver_issue_error_count_total",
```

Comment:
```suggestion
				Name:      "issue_errors_total",
```

If we want to keep `driver_` in the full metrics name, I suggest including it in the subsystem part above.

Naming is hard, but I tried to read https://prometheus.io/docs/practices/naming/, and I think count/total is almost the same.